6/map fragment

likppi10 opened the pull request:
### 1. 개요

화면 구성을 완료했고, 발생한 두가지 문제점 전부 해결했으나, 두번째 문제점에서 남은 의문점 및 문제점들이 있어 논의가 필요합니다. 

### 2. 작업사항
- 지도 화면, 건물 리스트 화면, 장소 검색 및 결과 화면 구현
- ca2d18562fd41f37d30a96792f63b4f2abf4e84e [MapView inflate 관련 문제]
- c8312ac96b2daed76533638f36db73ea0c040039 [Marker 관리 문제]

#### ca2d18562fd41f37d30a96792f63b4f2abf4e84e [MapView inflate 관련 문제] 
 : 카카오 맵 내부 정책으로 인하여 한 뷰에 2개 이상의 지도를 inflate 하는 것이 방지되어있습니다. 그리고 여기서 문제가 발생했습니다. 이전에 viewModel을 사용하지 않을 때는 문제가 없었으나, viewModel을 사용하기 위해 hilt를 적용하며 다음과 같은 문제가 발생했습니다. 

<img src="https://user-images.githubusercontent.com/54509842/183907178-a09cfae6-2dce-4b54-8f16-a3998f230271.png" width="800">
이 문제는 viewmodel을 MapFragment에 적용하며. xml에 net.daum.mf.map.api.MapView으로 카카오맵을 정의한 것이 중복적으로 inflate 것으로 간주되어 생긴 문제인 것으로 추측됩니다.

<img src="https://user-images.githubusercontent.com/54509842/183907947-9c0ae116-2ded-4fde-aa4e-7c0959dbd22e.png" width="800">
이를 해결하기 위해 xml에서 net.daum.mf.map.api.MapView으로 호출하고 있던 카카오 맵을 RelativeLayout으로 바꿔서 액티비티에서 inflate 해주는 것으로 중복적으로 inflate 되던 문제를 해결했습니다.

#### c8312ac96b2daed76533638f36db73ea0c040039 [Marker 관리 문제]
 : MapFragment에서 줌 레벨에 따라 마커를 생성, 삭제를 반복합니다. 이때 MapView에 마커 자체는 잘 관리가 되어지고 있다고 판단했습니다. 하지만, 다음과 같은 프로세스를 작동할 때 문제가 발생했습니다. 
1. 줌레벨을 확대했다가 축소하여 BuildingMarker와 Marker를 둘 다 화면에 띄워지나 확인합니다.
2. 검색 창을 눌러 검색화면으로 갔다가 나옵니다. 
3. 지도 화면으로 돌아오면 마커가 다음 그림처럼 비정상적으로 그려져 있습니다. 

<img src="https://user-images.githubusercontent.com/54509842/183909952-bc1757d5-7755-4dd7-b8bc-5e181fdefb96.png" width="300">

해서 MapFragment에서 다른 액티비티를 다녀올 때를 대비해서 마지막 줌 레벨(lastZoomLevel)과 마지막 위경도(currentLatitude, currentLongitude)를 저장합니다. 그리고 onResume에서 이곳으로, 이 줌레벨로 포커스를 이동하면, 해당 상황에 필요한 마커들을 리스너에서 생성할 수 있도록 구현해서 해결했습니다. 

<img src="https://user-images.githubusercontent.com/54509842/183910634-c3e7f1c1-249a-448a-b077-846ddb0ffa8b.png" width="400">
onResume 코드입니다. 

<img src="https://user-images.githubusercontent.com/54509842/183910652-2442bb76-f8ba-4095-bff5-3361f9a6a81e.png" width="400">
마지막 위경도를 저장합니다.

<img src="https://user-images.githubusercontent.com/54509842/183910745-25664218-3046-4928-ae1a-a8312bfb8ef6.png" width="400">
줌레벨 리스너에 마지막 줌레벨에 대한 조건을 추가해줍니다.

작동영상에서 확인해보시면 아실 수 있듯이 우리가 원하는 방향으로 잘 작동함을 알 수 있습니다.

하지만 다음과 같은 문제점이 있습니다. 

1. 검색화면에 갔다가 나왔을 때 다시 마커를 띄우는 시간이 다소 걸립니다. 
2. onResume에 이런 로직을 처리하는 것이 맞는지 다소 의문이 듭니다. (생명주기 상으로 액티비티에 갔다오면 다시 여기를 거치길래 우선 여기다가 구현해놓기는 했지만,,, 잘 모르겠군요)

### 3. 변경로직
1. 이번 브랜치에서는 xml closed column의 줄바꿈 처리가 적용됐습니다
3. 마커 네이밍에 관하여
- Marker : 줌 레벨에 따라 달라지는 마커이고 가장 최종 줌 레벨의 마커를 제외하면 데이터 폼이 같기 때문에 통용적으로 Marker라고 적용했습니다.
- BuildingMarker : 최종 줌 레벨의 건물 마커로, 이전과는 데이터 폼이 다르고, 건물을 지칭하는 불변할 마커이므로 BuildingMarker로 적용했습니다.

### 4. 스크린샷
https://user-images.githubusercontent.com/54509842/183905171-39f48578-7cb6-46eb-95ff-8c1b841cf213.mp4

### 5. 참고링크
- [디자인](https://www.figma.com/file/VBtuKLaxlF1ZaKEXCh505b/hometerview?node-id=162%3A8082)

### 6. 기타
++ 
1. 로직 개발 시에는 권한 관련 로직과 현위치 관련 로직이 추가, 수정되어야 합니다!
2. c8312ac96b2daed76533638f36db73ea0c040039 [Marker 관리 문제]를 해결하고 부각된 문제점에 대해 함께 논의해봐야 합니다. 
3. api 통신 시에는 현재 맵의 좌상단 우하단의 포인트를 넘겨주므로 좌표나 위치가 아닌 다음 사진의 MapPoint로 파라미터를 넘겨줘야합니다.
<img src="https://user-images.githubusercontent.com/54509842/183912996-8e33df29-06b2-4469-967e-ca7556046152.png" width="400">

likppi10 committed:
#6 map_fragment/지도 화면 : 카카오 맵 구성 파일 추가

likppi10 committed:
#6 map_fragment/지도 화면 : 카카오 맵 구성 이미지 추가

likppi10 committed:
#6 map_fragment/지도 화면 : 카카오 맵 구성 파일 추가

likppi10 committed:
#6 map_fragment/지도 화면 : 카카오 맵 구성 이미지 추가

likppi10 committed:
#6 map_fragment/지도 화면 : 카카오 맵 구현
- 카카오맵 띄우기
- stationMarker / buildingMarker 띄우기

likppi10 committed:
#6 map_fragment/지도 화면 : 카카오 맵 리스트 버튼, 현위치 버튼

likppi10 committed:
#6 refactor/xml : 블록 말미 줄바꿈 🔨

likppi10 committed:
#6 map_fragment/지도 화면 : 지도 화면
- 건물 리스트 화면
- 지하철 역, 건물명 검색 화면

likppi10 committed:
#7 Button/메인 버튼 :
- 메인 버튼 추가 : enrollCompanyNameFragment에서 사용함

likppi10 committed:
#6 map_fragment/지도 화면 : 카카오 맵 줌 레벨 설정

likppi10 committed:
PR Feedback : textAlignment 특정해서 적용

likppi10 committed:
PR Feedback : 애니메이션 끝나는 시점 감지

likppi10 committed:
PR Feedback : intent filter 이동

likppi10 committed:
Merge branch '6/rough_ui' of https://github.com/hometerview/Android into 6/map_fragment

 Conflicts:
	app/src/main/res/values/dimen.xml

likppi10 committed:
Merge branch '6/map_fragment' of https://github.com/hometerview/Android into 7/Button

 Conflicts:
	app/src/main/java/com/ftw/hometerview/ui/main/map/MapFragment.kt

likppi10 committed:
#7 Button/메인 버튼 :
- map_fragment에서 줌 레벨 관련 변경사항을 pull 해옵니다.

likppi10 committed:
Merge pull request #15 from hometerview/7/Button

#7 Button/메인 버튼 :

likppi10 committed:
Merge branch '6/map_fragment' of https://github.com/hometerview/Android into 6/map_fragment

 Conflicts:
	design/src/main/java/com/ftw/hometerview/design/MainButton.kt

likppi10 committed:
#6 map_fragment/지도 화면 : 메인 버튼 패키지 네임 수정

likppi10 committed:
#6 map_fragment/지도 화면 : 바인딩 선언 수정

likppi10 committed:
#6 map_fragment/지도 화면 : 빌딩 리스트 화면 구현
 - LocationReviewListViewModel 멀티 DI
 - 빌딩 리스트 화면 리스트 구현

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : domain 레이어 오타로 인한 리팩토링 🔨
 - entitiy > entity 로 이동
 - OnboardingResultViewModel > SearchCompanyResultViewModel로 rename
 - SearchResult > SearchCompanyResult로 rename

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : domain 레이어 오타로 인한 리팩토링 🔨
 - entitiy > entity 로 이동
 - OnboardingResultViewModel > SearchCompanyResultViewModel로 rename
 - SearchResult > SearchCompanyResult로 rename

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : 리팩토링 🔨
 - ActivityViewModelModule - provideSearchCompanyResultViewModel rename

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : 검색 결과 화면 📰
 - SearchAddressBuildingResult 추가

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : 검색 결과 화면 📰
 - GetSearchAddressBuildingUseCase(Impl) 추가

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : 검색 결과 화면 📰
 - SearchAddressBuildingUseCaseModule 추가

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : 검색 결과 화면 📰
 - SearchAddressBuildingUseCaseModule 추가

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : 검색 결과 화면 📰
 - SearchAddressBuilding을 위한 액티비티, ui, viewmodel 추가
 - 검색 버튼 클릭 시에 리사이클러뷰 생성

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : 검색 결과 화면 📰
 - 지역/지하철역/건물명 분기를 위한 바인딩 어댑터 등록

likppi10 committed:
#6 map_fragment/지역,지하철역,건물 검색 화면 : 리팩토링
 컨밴션 적용 (xml 줄바꿈, fragment binding 사용방법)

likppi10 committed:
#6 map_fragment/지도 화면 : 지도 화면
 - GetMarkerUseCase(Impl) 추가
 - GetBuildingMarkerUseCase(Impl) 추가
 - 모듈에 등록

likppi10 committed:
#6 map_fragment/지도 화면 : 지도 화면
 - MapViewModel 추가 및 모듈에 등록

likppi10 committed:
#6 map_fragment/지도 화면 : 지도 화면
 - MapViewModel을 MapFragment에서 사용하기 위해 MapView 사용 방식 변경 (inflate 관련 문제 해결)
 - viewmodel 사용

likppi10 committed:
#6 map_fragment/지도 화면 : 지도 화면
 - MapFragment에서 마커가 잔존하는 현상 해결 (Marker 관리 문제)

likppi10 committed:
Merge branch '6/rough_ui' of https://github.com/hometerview/Android into 6/map_fragment

 Conflicts:
	app/src/main/AndroidManifest.xml
	app/src/main/java/com/ftw/hometerview/di/ui/ActivityViewModelModule.kt
	app/src/main/res/layout/list_item_location_review.xml

likppi10 committed:
#6 map_fragment/pull :
- rough_ui에서 변경사항을 pull 해옵니다.

HayleyKim0716 commented on app/src/main/java/com/ftw/hometerview/bindingadapter/SearchResultAddressIconBindingAdapter.kt:
```diff
@@ -0,0 +1,39 @@
+import androidx.databinding.BindingAdapter
+import com.ftw.domain.entity.SearchAddressBuildingResult
+
+@BindingAdapter("android:icon")
+fun setSearchResultIcon(
+    textView: TextView,
+    item: SearchAddressBuildingResult
+) {
+    when (item.type) {
+        "지하철역" -> {
+            textView.text = "지하철역"
+        }
+        "지역" -> {
+            textView.text = "지역"
+        }
+        else -> {
+            textView.text = "건물"
+        }
+    }
+}
```
textView 의 BindingAdapter 로 정의된거지? 
음 아래 속성도 그렇고 general 하게 쓰긴 어려워 보이는데, viewModel 에 정의해서 사용하긴 어려울까?
위의 `android:icon` 도 textView 의 text 를 set 하는 것과 연관지어 생각하긴 어려운데, 어떤 관계가 있는걸까?.?

likppi10 replied:
address는 타입에 따라서 아이템 뷰의 주소를 가리키는 textView의 visibility를 GONE으로 바꿔줘야 해서,, 
일반적으로 사용할 수 없음에는 동의하지만, 뷰모델에서 이를 결정짓게 하는게 맞을까 하는 생각은 들어요!

icon은 텍스트랑 관련있는 것은 아니고, 지금 아이콘 디자인이 안 나와서 임의로 textView에 검색결과의 타입을 넣어줬어요! 
그래서 아마 textView가 아닌 imageView가 될 가능성이 클 것 같아요!

HayleyKim0716 commented on app/src/main/java/com/ftw/hometerview/ui/main/map/MapFragment.kt:
```diff
@@ -1,32 +1,345 @@
+        super.onResume()
+        if (::mapView.isInitialized && lastZoomLevel != ZoomLevel.NONE) {
+            mapView.removeAllPOIItems()
+            var zoom = 0
+            if (lastZoomLevel == ZoomLevel.LEVEL2) {
+                zoom = 4
+            } else if (lastZoomLevel == ZoomLevel.LEVEL1) {
+                zoom = 2
+            }
```
0, 4, 2 등이 어떤 의미인지 명확히 와 닿지는 않은데, enum class 로 의미를 명확히 하면 어떨까?

likppi10 replied:
ZoomLevel 안에 값을 넣어볼게요! 0, 4, 2 는 실제 줌레벨이에요!

HayleyKim0716 commented on app/src/main/java/com/ftw/hometerview/ui/main/map/MapFragment.kt:
```diff
@@ -1,32 +1,345 @@
+    lateinit var viewModel: MapViewModel
+    private var _binding: FragmentMapBinding? = null
+    private val binding get() = _binding!!
+    private lateinit var markerRootView: View
+    private lateinit var stationTextview: TextView
+    private lateinit var cntTextview: TextView
+    private lateinit var mapView: MapView
```
여기는 왜 binding 을 사용하지 않은거야~?,? 
혹시 view 가 여럿이라 그런거라면 다른 이름으로 binding 을 정의하는 건 어떨까?

likppi10 replied:
private lateinit var mapView: MapView

요거는 inflate 해 줄 지도라 따로 만들었어요 

위에 잇는 것들은 마커 때문에 만들었는데 binding을 따로 만들어볼게요!

HayleyKim0716 commented on app/src/main/java/com/ftw/hometerview/ui/searchaddressbuilding/SearchAddressBuildingActivity.kt:
```diff
@@ -0,0 +1,69 @@
+        initList()
+        observe()
+
+        binding.searchStationBuildingButton.setOnEditorActionListener { v, actionId, event ->
+            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
+                // 보내기 동작
+                viewModel.setLocation(binding.searchStationBuildingButton.text.toString())
+            }
+            true
+        }
+    }
```
요 부분도 BindingAdapter 로 정의해서 사용할 수 있을 것 같은데 어려울까?

likppi10 replied:
viewModel.setLocation() 이라는 함수를 바인딩 어댑터에서 불러다가 실행해 줄 수 있을까요? 그렇다면 가능할 것 같아요

HayleyKim0716 commented on app/src/main/java/com/ftw/hometerview/ui/buildinglist/BuildingListActivity.kt:
```diff
@@ -0,0 +1,77 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        binding = ActivityBuildingListBinding.inflate(layoutInflater)
+        setContentView(binding.root)
```
data binding 으로 viewModel 을 set 하는 부분은 어디있는걸까?.?

likppi10 replied:
DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)
            .apply {
                viewModel = this@MainActivity.viewModel
                lifecycleOwner = this@MainActivity
            }
액티비티에서 사용하는 이 코드에서 이렇게 코드를 짜 줘야만, xml에 있는 data에 정의된 뷰모델이 본 액티비티의 생명주기를 따를 것이다. 라고 설정한다고 이해하면 될까요?

HayleyKim0716 reviewed (approved):
고생 많았어 🙂
두 번째 문제는 왜 발생하는 걸까🥲
더 좋은 해결방법이 뭐가 있으려나,, 토요일에 같이 봐 볼까!?

likppi10 committed:
Merge branch '6/rough_ui' of https://github.com/hometerview/Android into 6/map_fragment

likppi10 committed:
PR Feedback : 줌레벨을 애매한 값으로 분기하지 않고 enum class 안에 직접 추가

likppi10 committed:
PR Feedback : 마커 뷰의 binding 설정

likppi10 committed:
PR Feedback : 마커 뷰의 binding 설정

likppi10 committed:
PR Feedback : viewModel 연결

likppi10 committed:
PR Feedback : editText two-way Binding 적용

likppi10 committed:
PR Feedback : 뷰의 일반적인 사용이 불가한 바인딩 어댑터 삭제 및 xml에서 처리

likppi10 committed:
PR Feedback :
- 마커의 공통된 요소 뽑아서 makeMarker() 생성
- 중복 사용되던 binding 필요할 때마다 호출
- 뷰 -> 비트맵 코드 따로 빼서 View의 Extension 함수로 생성
- 안 쓰는 코드 정리

likppi10 commented:
inline 함수를 사용해서 MapView.POIItemEventListener, MapView.MapViewEventListener의 구성 요소들을 빼내서 사용하기로 했던 부분의 구현이 더뎌지며 ,,, 실패했습니다. 
우선 그 외의 사항들은 PR:Feedback으로 커밋, 푸시했습니다. 
헬프.... 입니다...!

likppi10 committed:
#6 map_fragment/지도 화면 : 지도 화면
 - 건물 마커 디자인 변경 및 적용

likppi10 committed:
#6 map_fragment/지도 화면 : 지도 화면
 - 다른 탭 이동 시 줌레벨 초기화

likppi10 committed:
Merge branch '6/rough_ui' of https://github.com/hometerview/Android into 6/map_fragment

 Conflicts:
	app/src/main/res/layout/fragment_home.xml
	app/src/main/res/layout/list_item_location_review.xml
	app/src/main/res/values/dimen.xml
